tests/pipelines/test_train_step.py

```diff
@@ -157,3 +232,92 @@ def test_train_steps_with_correct_tags(tmp_pipeline_root_path):
+    ):
+        train_step_output_dir = setup_train_dataset(tmp_pipeline_root_path)
+        train_step = setup_train_step(tmp_pipeline_root_path, use_tuning=True)
+        train_step._run(str(train_step_output_dir))
+    assert (train_step_output_dir / "best_parameters.yaml").exists()
+
+    best_params_yaml = read_yaml(train_step_output_dir, "best_parameters.yaml")
+    assert "alpha" in best_params_yaml
+    assert "penalty" in best_params_yaml
+    assert "eta0" in best_params_yaml
+
+    run_id = open(train_step_output_dir / "run_id").read()
+    parent_run_params = MlflowClient().get_run(run_id).data.params
+    assert "alpha" in parent_run_params
+    assert "penalty" in parent_run_params
+    assert "eta0" in parent_run_params
+
+
+@pytest.mark.parametrize(
+    "with_hardcoded_params, expected_num_tuned, expected_num_hardcoded",
+    [(True, 3, 1), (False, 3, 0)],
+)
+def test_train_step_with_tuning_output_yaml_correct(
+    tmp_pipeline_root_path, with_hardcoded_params, expected_num_tuned, expected_num_hardcoded
+):
+    with mock.patch.dict(
+        os.environ, {_MLFLOW_PIPELINES_EXECUTION_DIRECTORY_ENV_VAR: str(tmp_pipeline_root_path)}
+    ):
+        train_step_output_dir = setup_train_dataset(tmp_pipeline_root_path)
+        train_step = setup_train_step(
+            tmp_pipeline_root_path, use_tuning=True, with_hardcoded_params=with_hardcoded_params
+        )
+        train_step._run(str(train_step_output_dir))
+    assert (train_step_output_dir / "best_parameters.yaml").exists()
+
+    with open(os.path.join(train_step_output_dir, "best_parameters.yaml")) as f:
+        lines = f.readlines()
+        assert lines[0] == "# tuned hyperparameters\n"
+        assert lines[expected_num_tuned + 2] == "# hardcoded parameters\n"
+        assert len(lines) == expected_num_tuned + expected_num_hardcoded + 3
+
+
+def test_train_step_with_tuning_child_runs_and_early_stop(tmp_pipeline_root_path):
+    with mock.patch.dict(
+        os.environ, {_MLFLOW_PIPELINES_EXECUTION_DIRECTORY_ENV_VAR: str(tmp_pipeline_root_path)}
+    ):
+        train_step_output_dir = setup_train_dataset(tmp_pipeline_root_path)
+        train_step = setup_train_step(tmp_pipeline_root_path, use_tuning=True)
+        train_step._run(str(train_step_output_dir))
+
+    with open(train_step_output_dir / "run_id") as f:
+        run_id = f.read()
+
+    run = MlflowClient().get_run(run_id)
+    child_runs = train_step._get_tuning_df(run, params=["alpha", "penalty", "eta0"])
+    assert len(child_runs) == 2
+    assert "params.alpha" in child_runs.columns
+    assert "params.penalty" in child_runs.columns
+    assert "params.eta0" in child_runs.columns
+
+    ordered_metrics = list(child_runs["root_mean_squared_error"])
+    assert ordered_metrics == sorted(ordered_metrics)
+
+
+@pytest.mark.skipif("hyperopt" not in sys.modules, reason="requires hyperopt to be installed")
+def test_search_space(tmp_pipeline_root_path):
```

Comment:
I think this test is always skipped. `hyperopt` is lazily imported and it doesn't exist in `sys.modules` when pytest collects tests.

Reply:
```
tests/pipelines/test_train_step.py::test_train_step_with_tuning_output_yaml_correct[False-3-0] PASSED [ 76%]
tests/pipelines/test_train_step.py::test_train_step_with_tuning_child_runs_and_early_stop PASSED [ 76%]
tests/pipelines/test_train_step.py::test_search_space SKIPPED (requi...) [ 77%]
tests/pipelines/test_train_step.py::test_tuning_param_equal[1-1] PASSED  [ 77%]
```

https://github.com/mlflow/mlflow/actions/runs/3139683814/jobs/5100333723#step:5:184

Reply:
Mark as a local only test?

Reply:
This test is skipped even on local.

Reply:
This test should be run on CI as well, right?

Reply:
Maybe we could just get rid of this line since `test-requirements.txt` contains `hyperopt`.

Reply:
Great catch, @harupy @BenWilson2 ! Removed the `skipif`.